feat(repository-json-schema): enhance getJsonSchema for navig props

samarpanB opened the pull request:
Enhance `getJsonSchema` to describe navigational properties

fix #2630

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

samarpanB commented:
@bajtos Can you please do a quick review and see if I am on right track ? Its a WIP PR. I still need to add new tests.

bajtos commented:
@samarpanB thank you for the pull request, the changes look reasonable so far. I think you will need to add code setting `targetsMany` property to all relational decorators, see e.g. https://github.com/strongloop/loopback-next/pull/2592/files#diff-64c2bbb6c63151367ef935def7eb54da

samarpanB commented on examples/log-extension/src/__tests__/acceptance/log.extension.acceptance.ts:
```diff
@@ -3,44 +3,43 @@
 describe('log extension acceptance test', () => {
   let app: LogApp;
-  let spy: SinonSpy;
+  let spy: SinonSpy<[(string | undefined)?], void>;
```
There was an error showing up while running tests. It needed explicit definition of type for the generic param. So I added it. Rest of the file changes are just reordering of imports done by vscode extension TSHero. I hope that's fine.

bajtos replied:
I think these new errors are reported by the newer version of TypeScript compiler that we recently upgraded to. Can you please rebase your patch on top of the latest master? I think the log-extension example should be already fixed by now.

/cc @raymondfeng

samarpanB replied:
I already did. Let me run the tests again and check.

samarpanB replied:
works. I removed my changes from here.

samarpanB commented on examples/log-extension/src/__tests__/unit/providers/log-action.provider.unit.ts:
```diff
@@ -3,23 +3,22 @@
 
 describe('LogActionProvider with in-memory logger', () => {
-  let spy: sinon.SinonSpy;
+  let spy: sinon.SinonSpy<[(string | undefined)?], void>;
```
There was an error showing up while running tests. It needed explicit definition of type for the generic param. So I added it. Rest of the file changes are just reordering of imports done by vscode extension TSHero. I hope that's fine.

samarpanB commented:
@bajtos I updated the PR with all the changes and test cases also added now. 

Just a note. I kind of struggled a lot this time to run the tests. 'npm test' kept on failing. First, it was due to missing deps. So, I ran 'npm ci'. But it also failed due to access issues. I am on a linux machine. After a long search/hit-and-trial, I figured that it was not working properly when I was using my own user in the machine. When I switched to root user, it worked perfectly. Do you think this is something we should document somewhere for contributors to save some time in troubleshooting ?

samarpanB commented:
@bajtos I am not sure why the build is failing. It is showing permission errors. I think that's unrelated.

bajtos commented:
> Just a note. I kind of struggled a lot this time to run the tests. 'npm test' kept on failing. First, it was due to missing deps. So, I ran 'npm ci'. 

Yes, you need to run `npm ci` to install dependencies before you can run `npm test`, see https://loopback.io/doc/en/contrib/code-contrib-lb4.html#building-the-project

> But it also failed due to access issues. I am on a linux machine. After a long search/hit-and-trial, I figured that it was not working properly when I was using my own user in the machine. When I switched to root user, it worked perfectly. Do you think this is something we should document somewhere for contributors to save some time in troubleshooting ?

I find that really weird, to be honest. Did you perhaps run `git clone` under root user? I am not aware of any dependencies that would require global install or root privileges. `npm ci` is supposed to install all dependencies locally, inside your checked-out repository.

Besides running `npm install`, we are also running `lerna` to create symlinks between monorepo packages. Is it perhaps possible that these symlinks cannot be created?

Anyhow, it's hard for me to help you without seeing the particular error logs. This is also very off topic in this pull request. Can you open a new issue please where we can take a closer look at the problems you are experiencing?

samarpanB commented:
> > Just a note. I kind of struggled a lot this time to run the tests. 'npm test' kept on failing. First, it was due to missing deps. So, I ran 'npm ci'.
> 
> Yes, you need to run `npm ci` to install dependencies before you can run `npm test`, see https://loopback.io/doc/en/contrib/code-contrib-lb4.html#building-the-project
> 
> > But it also failed due to access issues. I am on a linux machine. After a long search/hit-and-trial, I figured that it was not working properly when I was using my own user in the machine. When I switched to root user, it worked perfectly. Do you think this is something we should document somewhere for contributors to save some time in troubleshooting ?
> 
> I find that really weird, to be honest. Did you perhaps run `git clone` under root user? I am not aware of any dependencies that would require global install or root privileges. `npm ci` is supposed to install all dependencies locally, inside your checked-out repository.
> 
> Besides running `npm install`, we are also running `lerna` to create symlinks between monorepo packages. Is it perhaps possible that these symlinks cannot be created?
> 
> Anyhow, it's hard for me to help you without seeing the particular error logs. This is also very off topic in this pull request. Can you open a new issue please where we can take a closer look at the problems you are experiencing?
Yes I agree it's off-topic. I'll raise a separate issue if I encounter this again. I just wanted a heads up if I was doing something wrong. Thanks a lot for the direction @bajtos

bajtos commented on packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts:
```diff
@@ -717,5 +718,182 @@ describe('build-schema', () => {
+            items: {$ref: '#/definitions/ProductWithRelations'},
+          },
+        },
+        title: 'CategoryWithRelations',
```
Can you list the schema fields in the order `title`, `properties`, `definitions` please? I find it difficult to read `expectedScheme` when the order is reversed.

samarpanB replied:
I had to order it like that because the getJsonSchema method is returning the data in that order and if I don't keep it like that, the test fails. Is there any other way out here ?

bajtos replied:
I see. Let's not worry about this too much, we can keep the current form.

bajtos commented on packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts:
```diff
@@ -717,5 +718,182 @@ describe('build-schema', () => {
+      const jsonSchema = getJsonSchema(Category, {includeRelations: true});
+      expect(jsonSchema).to.deepEqual(expectedSchema);
+
+      // To check for case when there are no other properties than relational
```
Please move this check to a new test case. 

It's a best practice to keep one (logical) assertion per test.

samarpanB replied:
fixed.

nabdelgadir commented on packages/repository/src/relations/relation.types.ts:
```diff
@@ -22,6 +22,13 @@ export interface RelationDefinitionBase {
+   * false for relations with a single target (e.g. BelongsTo, HasOne).
+   * This property is needed by OpenAPI/JSON Schema generator.
+   */
+  targetsMany?: boolean;
```
I think that this should be a required property like in the issue description. What do you think?

samarpanB replied:
I kept it optional for backward compatibility. But if you think otherwise I can make this as required. Thoughts @bajtos ?

nabdelgadir replied:
That's a good point, let's wait for @bajtos' thoughts, but can you please still add it to the relation definitions? e.g.:

```ts
export interface HasManyDefinition extends RelationDefinitionBase {
  type: RelationType.hasMany;
  targetsMany: true;
```

samarpanB replied:
added to relation definitions.

bajtos replied:
> I kept it optional for backward compatibility. 

Good point!

I agree that making `targetsMany` a required property can break existing applications or 3rd-party extensions. On the other hand, I would be surprised if any such app or extension actually did exist.

Here is the important part: what is the expected behavior when `targetsMany` is not set? I don't think it's a good idea to default to `false`.

I am proposing the following:
- Add a comment starting with `TODO(semver-major)` explaining that we should make `targetsMany` required in LB5
- Modify the code reading `targetsMany` to throw a helpful error when `targetsMany` does not exist or is `undefined`.  The most important place to change is in `modelToJsonSchema`, but there may be other places to update too.

bajtos replied:
Preferably add a unit test to verify that `modeltoJsonSchema` can handle missing `targetsMany` and demonstrate what happens in such case.

samarpanB replied:
> Modify the code reading targetsMany to throw a helpful error when targetsMany does not exist or is undefined

Don't you think throwing error in this case would also mean we are making it backward incompatible ?

bajtos replied:
As I understand it, the error will be thrown only when `modeltoJsonSchema` is called with the new option `includeRelations`. Since this option did not exist before, there shouldn't be any existing code setting it to `true` and thus there shouldn't be any code affected by the new error being thrown.

samarpanB replied:
Hmmm. Got it. Let add some error handling and a test case to verify it as well. This makes sense.

samarpanB replied:
Done. Please check now.

nabdelgadir commented on packages/repository-json-schema/src/keys.ts:
```diff
@@ -6,10 +6,15 @@
 import {MetadataAccessor} from '@loopback/metadata';
 import {JSONSchema6 as JSONSchema} from 'json-schema';
 
+export const enum MODEL_TYPE_KEYS {
```
nice 👍

samarpanB replied:
thanks :)

nabdelgadir commented on packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts:
```diff
@@ -717,5 +718,182 @@ describe('build-schema', () => {
+        Category,
+      );
+      const jsonSchema = getJsonSchema(Category);
+      // Make sure its not pulling the withrelations key
```
Nitpick: `its` -> `it's`

samarpanB replied:
done

bajtos commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -203,6 +204,31 @@ export function metaToJsonProperty(meta: PropertyDefinition): JSONSchema {
+ * @param relMeta Relation metadata object
+ * @param targetRef Schema definition for the target model
+ */
+export function getPropDefForRelatedModel(
```
We prefer to use full names instead of abbreviations.

How about the following?

```ts
export function getNavigationalPropertyForRelation(
  relationMeta: RelationMetadata,
  targetSchema: JSONSchema
): JSONSchema {
  // ...
}
```

samarpanB replied:
Yeah. That's better. Changing it. Will rebase and squash as well.

nabdelgadir commented:
Thanks for the contribution @samarpanB! It's now landed 👏
